cmd/util/ledger/migrations/account_key_deduplication_encoder.go

```diff
@@ -0,0 +1,414 @@
+// - value (4 bytes): stored key index
+// NOTE:
+//   - If number of elements in a run-length group exceeds maxRunLengthInEncodedMappingGroup,
+//     a new group is created with remaining run-length and the same storedKeyIndex.
+//   - Consecutive groups are adjoining groups that run-length is 1 and value is increased by 1.
+//   - When consecutive group flag is on, run-length is number of consecutive groups, and value is the value of the first group.
+func (groups mappingGroups) Encode() []byte {
+	if len(groups) == 0 {
+		return nil
+	}
+
+	buf := make([]byte, 0, len(groups)*(mappingGroupSize))
+	off := 0
+
+	for i := 0; i < len(groups); {
+		group := groups[i]
+
+		switch group.runLength {
+		case 0:
+			panic(fmt.Sprintf("run length shouldn't be 0, mapping groups %+v", groups))
+
+		case 1:
+			// Handle consecutive groups
+
+			consecutiveGroupCount := uint32(1)
+			consecutiveStartStoredKeyIndex := group.storedKeyIndex
+
+			nextGroupIndex := i + 1
+			for nextGroupIndex < len(groups) {
+				nextGroup := groups[nextGroupIndex]
+				if nextGroup.runLength == 1 &&
+					nextGroup.storedKeyIndex == consecutiveStartStoredKeyIndex+consecutiveGroupCount {
+					consecutiveGroupCount++
+					nextGroupIndex++
+				} else {
+					break
+				}
+			}
+
+			if consecutiveGroupCount == 1 {
+				// Encode regular mapping group
+				buf, off = encodeMappingGroup(buf, off, group.runLength, group.storedKeyIndex, false)
+			} else {
+				// Encode consecutive mapping groups
+				buf, off = encodeMappingGroup(buf, off, consecutiveGroupCount, consecutiveStartStoredKeyIndex, true)
+			}
+
+			i = nextGroupIndex
+
+		default:
+			buf, off = encodeMappingGroup(buf, off, group.runLength, group.storedKeyIndex, false)
+			i++
+		}
+	}
+
+	return buf
+}
+
+func encodeMappingGroup(buf []byte, off int, runLength uint32, value uint32, isConsecutiveGroup bool) ([]byte, int) {
+	encodedRunLength := uint32(0)
+
+	for encodedRunLength < runLength {
+		// NOTE: if number of elements in a group exceeds maxMappingCountInGroup, a new group is created with the same value.
+
+		runLength := min(runLength-encodedRunLength, maxRunLengthInEncodedMappingGroup)
+
+		if cap(buf) >= off+mappingGroupSize {
+			buf = buf[:off+mappingGroupSize]
+		} else {
+			buf = append(buf, make([]byte, mappingGroupSize)...)
+		}
+
+		if isConsecutiveGroup {
+			binary.BigEndian.PutUint16(buf[off:], uint16(runLength)|consecutiveGroupFlagMask)
+		} else {
+			binary.BigEndian.PutUint16(buf[off:], uint16(runLength))
+		}
+		off += runLengthSize
+
+		binary.BigEndian.PutUint32(buf[off:], value)
+		off += storedKeyIndexSize
+
+		encodedRunLength += runLength
+	}
+
+	return buf, off
+}
+
+// encodeAccountPublicKeyMapping encodes keyIndexMappings into concatenated run-length groups.
+// Each run-length group is encoded as:
+// - length (2 bytes) with max length as 2<<15-1
+// - stored key index (4 bytes)
+func encodeAccountPublicKeyMapping(mapping []uint32) ([]byte, error) {
```

Comment:
Optional: would be great to walk through an example from end to end, such as `[KeyA, KeyB, KeyB, KeyC, KeyD, KeyB]`.